### PR TITLE
[Fixit] Provide fixit when converting expressions of type T to [T] or…

### PR DESCRIPTION
**Pull request by nkcsgexi**
This another fixit I found useful when helping developers migrate. When an expected argument is of type [T] or UnsafePointer<T>, however the developer feeds an expression of Type T, the fixit adds brackets to the expression.

#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

… UnsafePointer<T>.

### CODE REVIEW

**Comment by nkcsgexi** [comment]
@swift-ci Please test and merge

**Comment by gribozavr** [comment]
Isn't inserting just '&' simpler?

**Comment by nkcsgexi** [comment]
I think in that case we need separate paths for UnsafePointer and Array. Any advantage of using &?

**Comment by gribozavr** [comment]
> Any advantage of using &?

It does not allocate an array on the heap.

**Comment by nkcsgexi** [comment]
@swift-ci Please test and merge

**Comment by nkcsgexi** [comment]
@gribozavr Sounds reasonable. I have changed [] to & for the UnsafePointer case.

File: lib/Sema/CSDiag.cpp [hunk 1]
```diff
@@ -3477,6 +3477,32 @@ bool FailureDiagnosis::diagnoseContextualConversionError() {
...
+      for (Type arg : contextualType->getAllGenericArgs(scratch)) {
+        if (arg->isEqual(exprType)) {
+          diagnose(expr->getLoc(), diagID, exprType, contextualType).
+            fixItInsert(expr->getStartLoc(), "[").fixItInsert(
+              Lexer::getLocForEndOfToken(CS->TC.Context.SourceMgr,
+                                         expr->getEndLoc()), "]");
+          return true;
+        }
+      }
+    } else if (contextDecl == CS->TC.Context.getUnsafePointerDecl()) {
```
Comment (gribozavr) [inline lib/Sema/CSDiag.cpp]: What about `UnsafeMutablePointer`?

Comment (nkcsgexi) [inline lib/Sema/CSDiag.cpp]: Can we fix that in the same way?

Comment (gribozavr) [inline lib/Sema/CSDiag.cpp]: Yes.

**Comment by nkcsgexi** [comment]
@swift-ci Please test and merge

**Pull request merged**